sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -702,26 +707,39 @@ abstract class TypeCoercionBase {
         }.getOrElse(b)  // If there is no applicable conversion, leave expression unchanged.
 
       case e: ImplicitCastInputTypes if e.inputTypes.nonEmpty =>
-        val children: Seq[Expression] = e.children.zip(e.inputTypes).map { case (in, expected) =>
+        val childrenBeforeCollations: Seq[Expression] = e.children.zip(e.inputTypes).map {
           // If we cannot do the implicit cast, just use the original input.
-          implicitCast(in, expected).getOrElse(in)
+          case (in, expected) => implicitCast(in, expected).getOrElse(in)
+        }
+        val st = getOutputCollation(e.children)
+        val children: Seq[Expression] = childrenBeforeCollations.map {
```

Comment:
what are we doing here? `ImplicitCastInputTypes` defines the expected data type of each input, but does not require all inputs to be the same data type.

Reply:
Well the problem is that we have TypeCollections as expected types. If someone defines StringType as non-first DataType (and consequently non-defaultConcreteType), we need to go through previous elements first to see if it will come to StringType cast. If it does, we need to cast it to the same collation. For example, look at `ConcatWs` if it receives a BinaryType, it should try first to cast to ArrayType, get None and then try to cast it to StringType. But on the other hand, we need for consistency with findWiderCommonType to cast collations at the beginning of rules, as we would get a breakage in castings otherwise.

Reply:
It seems we are fixing the problem at the wrong place. If `TypeCollections` has issues, let's update the code in `def implicitCast` to fix it.

Reply:
Once again, this is the problem, we need to call getOutputCollation before we call implicitCast, as implicitCast only concerns one child, and not all children as a group. For collations we need to be sure we decided on the right collation first and then enter a casting rule. But if we do this, we would need to pass additional stuff to implicitCast, which changes the logic of the core function. Another idea was to implement implicitCasting for collations solely in a new rule, but we would end up copying a lot of already existent code in the implicitCast for TypeCollections and StringTypes.

Reply:
Note: if we want to have correlation between function inputs, please match the expression explicitly to do so. `ImplicitCastInputTypes` does not indicate input correlation and please DO NOT make such assumptions here.

Looking at `ConcatWs`, it does need the inputs to use the same string collation. Let's match it in the new rule and deal with it correctly.

In principle, let's not generalize things without seeing the full picture. What we should do to `ConcatWs` does not necessarily apply to all `ImplicitCastInputTypes` implementations.